stock_available_unreserved/models/product.py

```diff
@@ -0,0 +1,118 @@
+
+from odoo import api, fields, models, _
+from odoo.addons import decimal_precision as dp
+
+UNIT = dp.get_precision('Product Unit of Measure')
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    qty_available_not_res = fields.Float(
+        string='Quantity On Hand Unreserved',
+        digits=UNIT,
+        compute='_compute_product_available_not_res',
+    )
+
+    qty_available_stock_text = fields.Char(
+        compute='_compute_product_available_not_res',
+        string='Unreserved stock quantity',
+    )
+
+    @api.multi
+    @api.depends('product_variant_ids.qty_available_not_res')
+    def _compute_product_available_not_res(self):
+        no_new = self.filtered(lambda x: not isinstance(x.id, models.NewId))
+        for tmpl in no_new:
+            tmpl.qty_available_not_res = sum(
+                tmpl.mapped('product_variant_ids.qty_available_not_res')
+            )
+            tmpl.qty_available_stock_text = "/".join(
+                tmpl.mapped('product_variant_ids.qty_available_stock_text')
+            )
+
+    @api.multi
+    def action_open_quants_unreserved(self):
+        products = self.mapped('product_variant_ids').ids
+        quants = self.env['stock.quant'].search(
+            [
+                ('product_id', 'in', products),
+            ]
+        )
+        quant_ids = quants.filtered(
+            lambda x: x.product_id.qty_available_not_res > 0
+        ).ids
+        result = self.env.ref('stock.product_open_quants').read()[0]
+        result['domain'] = "[('id', 'in', {})]".format(quant_ids)
+        result['context'] = "{'search_default_locationgroup': 1, " \
+                            "'search_default_internal_loc': 1}"
+        return result
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    qty_available_not_res = fields.Float(
+        string='Qty Available Not Reserved',
+        digits=UNIT,
+        compute='_compute_qty_available_not_reserved',
+    )
+
+    qty_available_stock_text = fields.Char(
+        compute='_compute_qty_available_not_reserved',
+        string='Available per stock',
+    )
+
+    @api.multi
+    def _product_available_not_res_hook(self, quants):
+        """Hook used to introduce possible variations"""
+        return False
+
+    @api.multi
+    def _prepare_domain_available_not_reserved(self):
+        domain_quant = [
+            ('product_id', 'in', self.ids),
+            ('contains_unreserved', '=', True),
+        ]
+        domain_quant_locations = self._get_domain_locations()[0]
+        domain_quant.extend(domain_quant_locations)
+        return domain_quant
+
+    @api.multi
+    def _compute_product_available_not_res_dict(self):
+
+        res = {}
+
+        domain_quant = self._prepare_domain_available_not_reserved()
+        quants = self.env['stock.quant'].with_context(lang=False).search(
+            domain_quant,
+        )
+        for prod in self:
```

Comment:
just a note for future perf-wise refactoring: wouldn't be better to pick up the quants already grouped by product.
This multiple loop + filtered sounds sub-optimal when you have tons of products and quants.
Would you mind adding a comment like `'TODO: rething this for better performance (maybe using a raw query)`?

Reply:
@simahawk this actually requires further testing and optimisation.
as:
1. This is used only in compute field, so i don't think it will hold a ton of product every time
2. Seems that there can no longer be several quants for one product on same location. And they are basically per product per location. 
I mimicked the behaviour here, but a review from more knowledgable person in stock BP is required :)

Reply:
I don't see a constraint on the quant model. Thus it is not safe to assume uniqueness of quants on product + location

You can have a look at the stock module how to perfectly mimic the count using a read_group.

https://github.com/odoo/odoo/blob/d28f8f704febe030f9c068f895adf0b1ce3e8dc1/addons/stock/models/product.py#L125

Reply:
Doesn't need to be part of this pull request, this is a migration, not a refactoring :) But as @simahawk says a note would be worth to add

Reply:
I missed the fact that in 10.0 it was a `read_group`

Reply:
So, as we understand, would be:


....
domain_quant = self._prepare_domain_available_not_reserved()
quants_res = self.env['stock.quant'].with_context(lang=False).read_group(domain_quant, ['product_id', 'quantity', 'reserved_quantity'], ['product_id'], orderby='id')
for prod in self:
    quantity = sum(datum['quantity'] - datum['reserved_quantity'] for 
    datum in quants_res)
    vals['qty_available_not_res'] = quantity
    res[prod.id] = vals
....